content/en/blog/_posts/2024-09-25-swap-ga-graduation.md

```diff
@@ -0,0 +1,475 @@
+    Min-->>+High: Back to normal.
+{{< /mermaid >}}
+
+To figure out what's the value of the different watermarks, one can do the following (the output is trimmed
+for simplicity. values are in memory pages):
+```shell
+> cat /proc/zoneinfo
+--
+Node 0, zone    DMA32
+        min      569
+        low      1120
+        high     1671
+--
+Node 0, zone   Normal
+        min      16322
+        low      32138
+        high     47954
+--
+...
+```
+
+This is a real example from a 64GB RAM machine with the default kernel configurations. On such a machine:
+* The `high` watermark is at `187.32Mi`.
+* The `low` watermark is at `125.54Mi`.
+* The `min` watermark is at `63.76Mi`.
+
+An advanced cluster-admin can choose to set kubelet's memory eviction threshold according to these values on nodes
+with swap enabled.
+For example, the threshold can be set to a value between the low and min watermarks.
+Bear in mind that the default memory eviction threshold of `100Mi` is between the low and min watermarks as well,
+which means that by default evictions are expected to work decently.
+
+## Good practice for using swap in a Kubernetes cluster
+
+### Disable swap for system-critical daemons
+
+During the testing phase and based on user feedback, it was observed that the performance
+of system-critical daemons and services might degrade.
+This implies that system daemons, including the kubelet, could operate slower than usual.
+If this issue is encountered, it is advisable to configure the cgroup of the system slice
+to prevent swapping (i.e., set `memory.swap.max=0`).
+
+### Protect system-critical daemons for I/O latency
+
+Swap can increase the I/O load on a node.
+When memory pressure causes the kernel to rapidly swap pages in and out,
+system-critical daemons and services that rely on I/O operations may
+experience performance degradation.
+
+To mitigate this, it is recommended to prioritize the system slice in terms of I/O latency.
+This can be achieved by setting `io.latency` for the system slice,
+thereby granting it higher I/O priority.
+
+### Swap and control plane nodes
+
+The Kubernetes project recommends running control plane nodes without any swap space configured.
+The control plane primarily hosts Guaranteed QoS Pods, so swap can generally be disabled.
+The main concern is that swapping critical services on the control plane could negatively impact performance.
+
+### Use of a dedicated disk for swap
+
+It is recommended to use a separate, encrypted disk for the swap partition.
+If swap resides on a partition or the root filesystem, workloads may interfere
+with system processes that need to write to disk.
+When they share the same disk, processes can overwhelm swap,
+disrupting the I/O of kubelet, container runtime, and systemd, which would impact other workloads.
+Since swap space is located on a disk, it is crucial to ensure the disk is fast enough for the intended use cases.
+
```

Comment:
Alternative: configure I/O priorities between different mapped areas of one backing device.